### PR TITLE
Feat/local paymaster

### PR DESCRIPTION
**Pull request by Keshavrajsinghal**
**What changed? Why?**

In this PR, paymaster support for users registering basenames with smart wallets has been added. This includes a proxy that verifies whether the user is eligible for paymaster sponsorship as well as configuration that supports our paymaster setup.

**Notes to reviewers**

The array of valid networks has been edited where support for base and mainnet has been eliminated. This is only for testing purposes. With the current version of Coinbase smart wallets, we're unable to switch networks via the wallet which prohibits us from testing whether the sign up flow using a smart wallet works since our contracts are only deployed on baseSepolia for now.

**How has it been tested?**

Transaction flows with undeployed accounts, EoAs, and Coinbase Smart Wallets have been completed. These include minting a basename using different wallet types, checking for these transactions on basescan, and verifying the paymaster usage using the coinbase developer portal.

### CODE REVIEW

File: apps/web/pages/api/paymaster/index.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { paymasterClient } from "../../../src/utils/paymasterConfig";
+import { willSponsor } from "../../../src/utils/paymasterSponsor";
+
+export async function POST(r: Request) {
+    console.log('request received');
```
Comment (JFrankfurt) [inline apps/web/pages/api/paymaster/index.ts:5]:
```suggestion
```

File: apps/web/pages/_app.tsx [hunk 2]
```diff
@@ -1,12 +1,22 @@
 import './global.css';
 
+import '@rainbow-me/rainbowkit/styles.css';
+import { RainbowKitProvider } from '@rainbow-me/rainbowkit';
+import { WagmiProvider } from 'wagmi';
+import { base, mainnet, baseSepolia, sepolia } from 'wagmi/chains';
+import { QueryClientProvider, QueryClient } from '@tanstack/react-query';
+
```
Comment (JFrankfurt) [inline apps/web/pages/_app.tsx:8]:
```suggestion
```

File: apps/web/pages/_app.tsx [hunk 3]
```diff
@@ -15,20 +25,18 @@ import {
...
-import useSprig from 'base-ui/hooks/useSprig';
-import { MotionConfig } from 'framer-motion';
-import { AppProps } from 'next/app';
-import { ReactElement, ReactNode, useCallback, useEffect, useRef, useState } from 'react';
-import { createConfig, http, WagmiProvider } from 'wagmi';
-import { base, baseSepolia, mainnet, sepolia } from 'wagmi/chains';
-import ClientAnalyticsScript from '../src/components/ClientAnalyticsScript/ClientAnalyticsScript';
+
+import { createConfig, http } from 'wagmi';
+
```
Comment (JFrankfurt) [inline apps/web/pages/_app.tsx:30]:
```suggestion
```

File: apps/web/pages/_app.tsx [hunk 4]
```diff
@@ -15,20 +25,18 @@ import {
...
 import { Layout, NavigationType } from '../src/components/Layout/Layout';
+import ClientAnalyticsScript from '../src/components/ClientAnalyticsScript/ClientAnalyticsScript';
+
```
Comment (JFrankfurt) [inline apps/web/pages/_app.tsx:33]:
```suggestion
import { createConfig, http } from 'wagmi';
import { Layout, NavigationType } from '../src/components/Layout/Layout';
import ClientAnalyticsScript from '../src/components/ClientAnalyticsScript/ClientAnalyticsScript';
```

File: apps/web/pages/_app.tsx [hunk 5]
```diff
@@ -49,7 +57,7 @@ const connectors = connectorsForWallets(
 
 const config = createConfig({
   connectors,
-  chains: [mainnet, base, sepolia, baseSepolia],
+  chains: [mainnet, baseSepolia, sepolia, base],
```
Comment (JFrankfurt) [inline apps/web/pages/_app.tsx:60]: does the order change do anything here?

File: apps/web/pages/_app.tsx [hunk 6]
```diff
@@ -15,20 +25,18 @@ import {
...
+import ClientAnalyticsScript from '../src/components/ClientAnalyticsScript/ClientAnalyticsScript';
+
 import { cookieManagerConfig } from '../src/utils/cookieManagerConfig';
-import { NextPage } from 'next';
+import useSprig from 'base-ui/hooks/useSprig';
 import { UserAvatar } from 'apps/web/src/components/ConnectWalletButton/UserAvatar';
+import { NextPage } from 'next';
 
-coinbaseWallet.preference = 'all';
+coinbaseWallet.preference = 'smartWalletOnly';
```
Comment (JFrankfurt) [inline apps/web/pages/_app.tsx:39]: iirc `smartWalletOnly` is a dev setting and we want `all` here

File: apps/web/pages/_app.tsx [hunk 7]
```diff
@@ -58,11 +66,10 @@ const config = createConfig({
   },
   ssr: true,
 });
-
 const queryClient = new QueryClient();
 const sprigEnvironmentId = process.env.NEXT_PUBLIC_SPRIG_ENVIRONMENT_ID;
 
-export type NextPageWithLayout<P = unknown, IP = P> = NextPage<P, IP> & {
+export type NextPageWithLayout<P = object, IP = P> = NextPage<P, IP> & {
```
Comment (JFrankfurt) [inline apps/web/pages/_app.tsx:72]: is this change necessary? (TS usually complains about `object`)

File: apps/web/src/constants.ts [hunk 8]
```diff
@@ -30,5 +30,47 @@ export const ATTESTATION_VERIFIED_CB1_ACCOUNT_SCHEMA_IDS: AddressMap = {
...
+            type: "bytes",
+            internalType: "bytes",
+          },
+        ],
+      },
+    ],
+    outputs: [],
+    stateMutability: "payable",
+  },
+];
```
Comment (JFrankfurt) [inline apps/web/src/constants.ts:74]:
```suggestion
] as const
```

File: apps/web/src/utils/paymasterConfig.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  const client = createPublicClient({
+    chain: baseSepolia,
+    transport: http(),
+  });
+  
+  return client;
+}
+
+
+const paymasterService = 'https://api.developer.coinbase.com/rpc/v1/base-sepolia/1IhTcPOmhK5aEq-4WqRZMJoOh0oPenD2'; // Change this to the official paymaster
```
Comment (JFrankfurt) [inline apps/web/src/utils/paymasterConfig.ts:19]: is this comment still relevant?

File: apps/web/src/utils/paymasterConfig.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+export const paymasterClient = createClient({
+  chain: baseSepolia,
```
Comment (JFrankfurt) [inline apps/web/src/utils/paymasterConfig.ts:22]: should this be environment aware?

File: apps/web/src/utils/paymasterConfig.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+import { ENTRYPOINT_ADDRESS_V06 } from "permissionless";
+import { createClient, createPublicClient, http } from "viem";
+import { paymasterActionsEip7677 } from "permissionless/experimental";
+
+
+
+export function initializeClient() {
+  
+  const client = createPublicClient({
+    chain: baseSepolia,
```
Comment (JFrankfurt) [inline apps/web/src/utils/paymasterConfig.ts:11]: should this be environment aware?

File: apps/web/src/utils/paymasterSponsor.ts [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+  Address,
+  BlockTag,
+  Hex,
+  decodeAbiParameters,
+  decodeFunctionData,
+} from "viem";
+import { baseSepolia } from "viem/chains";
+import {initializeClient} from "../../src/utils/paymasterConfig";
+
+
```
Comment (JFrankfurt) [inline apps/web/src/utils/paymasterSponsor.ts:12]:
```suggestion
```

File: apps/web/src/utils/paymasterSponsor.ts [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    try {
+        const publicClient = initializeClient();
+        const code = await publicClient.getCode({ address: userOp.sender });
+     
+        if (!code) {
+          // no code at address, check that the initCode is deploying a Coinbase Smart Wallet
+          // factory address is first 20 bytes of initCode after '0x'
+          const factoryAddress = userOp.initCode.slice(0, 42);
+          if (factoryAddress.toLowerCase() !== CB_SW_FACTORY_ADDRESS.toLowerCase())
+            return false;
```
Comment (JFrankfurt) [inline apps/web/src/utils/paymasterSponsor.ts:41]:
```suggestion
          if (factoryAddress.toLowerCase() !== CB_SW_FACTORY_ADDRESS.toLowerCase()) {
            return false;
          }
```

File: apps/web/src/utils/paymasterSponsor.ts [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    if (entrypoint.toLowerCase() !== ENTRYPOINT_ADDRESS_V06.toLowerCase()) return false;
+
+    try {
+        const publicClient = initializeClient();
+        const code = await publicClient.getCode({ address: userOp.sender });
+     
+        if (!code) {
+          // no code at address, check that the initCode is deploying a Coinbase Smart Wallet
+          // factory address is first 20 bytes of initCode after '0x'
+          const factoryAddress = userOp.initCode.slice(0, 42);
```
Comment (JFrankfurt) [inline apps/web/src/utils/paymasterSponsor.ts:39]: double checking this is doing what you want, since the `0` is going to leave `0x` in the result

File: apps/web/pages/_app.tsx (hunk 7 above)
Comment (Keshavrajsinghal) [reply apps/web/pages/_app.tsx:72]: It complained about keeping it {}. Will change it to unknown

File: apps/web/src/utils/paymasterSponsor.ts [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    });
+    if (innerCalldata.functionName !== "register") return false;
+ 
+    return true;
+  } catch (e) {
+    console.error(`willSponsor check failed: ${e}`);
+    return false;
+  }
+}
+
```
Comment (JFrankfurt) [inline apps/web/src/utils/paymasterSponsor.ts:104]: could you update the three multi-line/no-braces if/return blocks to be single-line or use curlies?

File: apps/web/package.json [hunk 16]
```diff
@@ -16,7 +16,7 @@
     "@headlessui/react": "^1.7.19",
     "@heroicons/react": "^2.1.3",
     "@next/font": "^13.1.5",
-    "@rainbow-me/rainbowkit": "^2.1.2",
+    "@rainbow-me/rainbowkit": "2.1.3",
```
Comment (JFrankfurt) [inline apps/web/package.json:19]: do these need to be pinned, or can we keep the `^`?

File: apps/web/package.json [hunk 17]
```diff
@@ -42,7 +43,7 @@
     "usehooks-ts": "^3.1.0",
     "uuidv4": "^6.2.13",
     "viem": "2.x",
-    "wagmi": "^2.9.11"
+    "wagmi": "2.11.3"
```
Comment (JFrankfurt) [inline apps/web/package.json:46]: same q

File: apps/web/src/hooks/useRegisterNameCallback.ts [hunk 18]
```diff
@@ -22,10 +27,46 @@ export function useRegisterNameCallback(
...
+  
+  const capabilities = useMemo(() => {
+    if (!account.isConnected || !chainId || !availableCapacities) {
+      return {};
+    }
+    const chainCapabilities = availableCapacities[chainId];
+    if (chainCapabilities["paymasterService"] && chainCapabilities["paymasterService"].supported) {
+      return {
+        paymasterService: {
+          // url: `${document.location.origin}/api/paymaster`
```
Comment (JFrankfurt) [inline apps/web/src/hooks/useRegisterNameCallback.ts:56]: what's the significance of this comment?

File: apps/web/src/hooks/useRegisterNameCallback.ts [hunk 19]
```diff
@@ -22,10 +27,46 @@ export function useRegisterNameCallback(
...
+    }
+    return {};
+  }, [availableCapacities, chainId]);  
 
+  if (chainId === undefined) {
+    console.error(
+      'useRegisterNameCallback: Unable to create name registration callback; chainId must be defined.',
+    );
+    return () => {};
+  }
```
Comment (JFrankfurt) [inline apps/web/src/hooks/useRegisterNameCallback.ts:69]: the early return here will cause client side crashes if the code ever runs because this hook will be running a variable number of hooks

**Comment by JFrankfurt** [comment]
Have you tested the full discounted and undiscounted flows with smart wallet and a traditional injected wallet? (talk to @stevieraykatz if you need help getting discounts working)

File: apps/web/pages/api/paymaster/index.ts (hunk 1 above)
Comment (Keshavrajsinghal) [reply apps/web/pages/api/paymaster/index.ts:5]: Acknowledged and resolved

File: apps/web/pages/_app.tsx (hunk 5 above)
Comment (Keshavrajsinghal) [reply apps/web/pages/_app.tsx:60]: I don't believe it does, changed it back to the original format

File: apps/web/src/utils/paymasterConfig.ts (hunk 9 above)
Comment (Keshavrajsinghal) [reply apps/web/src/utils/paymasterConfig.ts:19]: Yes, will update this when the paymaster team sets up the official paymaster for this project. Using this url only for testing.

File: apps/web/src/utils/paymasterSponsor.ts (hunk 14 above)
Comment (Keshavrajsinghal) [reply apps/web/src/utils/paymasterSponsor.ts:39]: Yes, the factory address is configured that way as well.

File: apps/web/package.json (hunk 16 above)
Comment (Keshavrajsinghal) [reply apps/web/package.json:19]: I think we can keep the ^. I don't know what components are reliant on earlier versions but smart wallet support shouldn't be affected by earlier versions.

File: apps/web/package.json (hunk 17 above)
Comment (Keshavrajsinghal) [reply apps/web/package.json:46]: Yep, can keep the ^. Same reason.

File: apps/web/src/hooks/useRegisterNameCallback.ts (hunk 18 above)
Comment (Keshavrajsinghal) [reply apps/web/src/hooks/useRegisterNameCallback.ts:56]: Might have to uncomment that before deployment.

File: apps/web/src/utils/paymasterConfig.ts (hunk 9 above)
Comment (JFrankfurt) [reply apps/web/src/utils/paymasterConfig.ts]: can you make sure that we have a task in Linear to address anything like this (and reference the task # in the comment as well)

File: apps/web/src/hooks/useRegisterNameCallback.ts (hunk 18 above)
Comment (JFrankfurt) [reply apps/web/src/hooks/useRegisterNameCallback.ts:56]: if we need to do anything before production to ensure this works, make sure it's documented sufficiently (i.e., punchlist item, linear, etc. wherever it's relevant)
